### PR TITLE
feat(extract): add scoped call graph with sync connection detection (#237)

### LINKED ISSUE #237: [M1-D1.1/D1.2/D1.3/D1.4] Build scoped call graph with sync connection detection
### Context
**PRD:** \`docs/project/PRD/active/PRD-phase-12-connection-detection.md\` — M1 D1.1, D1.2, D1.3, D1.4

Connection detection requires building a method-level call graph from known components, tracing through non-components, resolving single-implementation interfaces, and detecting sync connections via method calls on component-typed instances. These four deliverables share the same state dependencies (ts-morph Project, EnrichedComponent[], ComponentIndex) and form a single vertical slice through the connection detection capability.

### Traceability
- **PRD Section:** M1-D1.1 (Scoped call graph), M1-D1.2 (Non-component transparency), M1-D1.3 (Interface resolution), M1-D1.4 (Sync connection detection)
- **Functional Requirements:** Type-based method call resolution, component-to-component edge production, single-impl interface auto-resolve
- **Success Criteria:** SC-1 (Golden Path sync connections), SC-3 (non-component tracing), SC-4 (interface resolution)

### Acceptance Criteria
- ✓ ComponentIndex value object built from EnrichedComponent[] + AST, supporting type-name lookup and file:line location lookup for method-level components
- ✓ Method-level call graph constructed via DFS from component methods using type-based resolution (declared types on constructor params, fields, variables)
- ✓ Cycle detection via visited set per traversal path — already-visited methods skipped
- ✓ Non-component classes traced through transparently: A → non1 → non2 → B produces A → B link
- ✓ Dead-end chains (component → non-component → nothing) produce no link
- ✓ One link per unique (source component, target component, type) tuple
- ✓ Source location for transitive connections references call site in source component (lexically first occurrence)
- ✓ Single-implementation interface/abstract class auto-resolves within extraction config module globs (node_modules excluded)
- ✓ Zero implementations: fail fast in strict mode, mark uncertain in lenient mode
- ✓ Multiple implementations: fail fast in strict mode, mark uncertain in lenient mode
- ✓ Sync connections detected when calling methods on instances whose declared type is a known component
- ✓ Constructor parameters provide type info — method call creates the link, not constructor declaration
- ✓ Every connection includes sourceLocation (file path, line number, method name)
- ✓ Unresolvable types: strict mode fails with error (file, line, type name, reason); lenient mode emits _uncertain
- ✓ Generic types match against base type (Repository<Order> matches component Repository)
- ✓ Self-links (source === target) excluded
- ✓ Getters traced into (have method bodies); property access without getter not traced
- ✓ Chained calls a.b().c() — each resolved independently via declared return type
- ✓ await is transparent
- ✓ 100% branch coverage

### Edge Case Scenarios
- Empty EnrichedComponent[] → empty links array, no errors
- Component with no method calls → no links produced
- Component calling only non-components that lead to dead ends → no links
- Circular call chain (A → B → A) → cycle detected, no infinite loop, produces A → B and B → A links
- Multiple call sites between same component pair → single link, lexically first source location
- Interface with zero implementations in strict mode → structured error with file/line/interface name
- Interface with 3+ implementations in strict mode → structured error listing implementation count
- Generic type Repository<Order> → matches component named Repository
- Method call on untyped variable (any) → strict fails, lenient marks uncertain
- Component method calling this.field where field type is another component → link produced
- Deeply nested non-component chain (5+ hops) → correctly traces through all
- Self-referential component (calls own methods) → no self-link produced

### Implementation Guidelines
- **Architecture:** New \`domain/connection-detection/\` capability in \`riviere-extract-ts\` per §9.1.2
  - \`call-graph/\` — Scoped call graph construction + DFS traversal
  - \`sync-detection/\` — Method call detection
  - \`interface-resolution/\` — Single-impl auto-resolve
  - \`detect-connections.ts\` — Orchestrator
- **Firm constraints:** §9.1.1 (no new packages), §9.1.2 (domain capability in extract-ts), §9.2.1 (ComponentIndex immutable value object)
- **Flexible decisions:** §9.2.2 (ScopedCallGraph implementation details), §9.2.3 (ExtractedLink extends Link)
- Use ts-morph for AST traversal (already a dependency of extract-ts)
- Follow existing patterns in \`domain/component-extraction/extractor.ts\` and \`domain/value-extraction/\`
- ExtractedLink extends Link from riviere-schema with optional \`_uncertain?: string\`
- Apply SD-004 (dependency injection), SD-001 (fail-fast), SD-010 (wrap primitives as value objects)
- Follow \`docs/conventions/codebase-structure.md\` CS-004 (organize by usage)

### Embedded Reasoning
**Why:** Components are identified and enriched. Detecting how they connect via method calls is the foundation of flow-based architecture extraction. Call graph + sync detection is the core capability everything else builds on.
**What:** ComponentIndex (immutable lookup), ScopedCallGraph (DFS traversal with type-based resolution), interface resolution (single-impl auto-resolve), sync connection detection (method calls → ExtractedLink[]).
**How:** Build ComponentIndex from EnrichedComponent[] + AST. For each component method, DFS through call sites resolving receiver types via declared types. Collapse non-component nodes. Produce one ExtractedLink per unique (source, target, type) tuple.

### Testing Strategy
- **Unit:** ComponentIndex construction and lookups (type-name, file:line). Call graph DFS traversal. Interface resolution (0, 1, N implementations). Sync detection for direct calls, transitive calls, chained calls.
- **Integration:** End-to-end from EnrichedComponent[] + ts-morph Project to ExtractedLink[] using multi-file test fixtures.
- **Edge Cases:** Empty inputs, cycles, dead ends, self-links, generic types, untyped variables, getters vs property access, await transparency.
- **Performance:** Not in scope for this task (D1.8).

### Dependencies
None — this is the foundational task.

### Verification
\`\`\`bash
pnpm nx test riviere-extract-ts -- --testNamePattern "connection-detection"
pnpm nx test riviere-extract-ts -- --coverage
\`\`\`

### PR DESCRIPTION
**Pull request by NTCoding**
Closes #237

### Context
**PRD:** \`docs/project/PRD/active/PRD-phase-12-connection-detection.md\` — M1 D1.1, D1.2, D1.3, D1.4

Connection detection requires building a method-level call graph from known components, tracing through non-components, resolving single-implementation interfaces, and detecting sync connections via method calls on component-typed instances. These four deliverables share the same state dependencies (ts-morph Project, EnrichedComponent[], ComponentIndex) and form a single vertical slice through the connection detection capability.

### Traceability
- **PRD Section:** M1-D1.1 (Scoped call graph), M1-D1.2 (Non-component transparency), M1-D1.3 (Interface resolution), M1-D1.4 (Sync connection detection)
- **Functional Requirements:** Type-based method call resolution, component-to-component edge production, single-impl interface auto-resolve
- **Success Criteria:** SC-1 (Golden Path sync connections), SC-3 (non-component tracing), SC-4 (interface resolution)

### Acceptance Criteria
- ✓ ComponentIndex value object built from EnrichedComponent[] + AST, supporting type-name lookup and file:line location lookup for method-level components
- ✓ Method-level call graph constructed via DFS from component methods using type-based resolution (declared types on constructor params, fields, variables)
- ✓ Cycle detection via visited set per traversal path — already-visited methods skipped
- ✓ Non-component classes traced through transparently: A → non1 → non2 → B produces A → B link
- ✓ Dead-end chains (component → non-component → nothing) produce no link
- ✓ One link per unique (source component, target component, type) tuple
- ✓ Source location for transitive connections references call site in source component (lexically first occurrence)
- ✓ Single-implementation interface/abstract class auto-resolves within extraction config module globs (node_modules excluded)
- ✓ Zero implementations: fail fast in strict mode, mark uncertain in lenient mode
- ✓ Multiple implementations: fail fast in strict mode, mark uncertain in lenient mode
- ✓ Sync connections detected when calling methods on instances whose declared type is a known component
- ✓ Constructor parameters provide type info — method call creates the link, not constructor declaration
- ✓ Every connection includes sourceLocation (file path, line number, method name)
- ✓ Unresolvable types: strict mode fails with error (file, line, type name, reason); lenient mode emits _uncertain
- ✓ Generic types match against base type (Repository<Order> matches component Repository)
- ✓ Self-links (source === target) excluded
- ✓ Getters traced into (have method bodies); property access without getter not traced
- ✓ Chained calls a.b().c() — each resolved independently via declared return type
- ✓ await is transparent
- ✓ 100% branch coverage

### Edge Case Scenarios
- Empty EnrichedComponent[] → empty links array, no errors
- Component with no method calls → no links produced
- Component calling only non-components that lead to dead ends → no links
- Circular call chain (A → B → A) → cycle detected, no infinite loop, produces A → B and B → A links
- Multiple call sites between same component pair → single link, lexically first source location
- Interface with zero implementations in strict mode → structured error with file/line/interface name
- Interface with 3+ implementations in strict mode → structured error listing implementation count
- Generic type Repository<Order> → matches component named Repository
- Method call on untyped variable (any) → strict fails, lenient marks uncertain
- Component method calling this.field where field type is another component → link produced
- Deeply nested non-component chain (5+ hops) → correctly traces through all
- Self-referential component (calls own methods) → no self-link produced

### Implementation Guidelines
- **Architecture:** New \`domain/connection-detection/\` capability in \`riviere-extract-ts\` per §9.1.2
  - \`call-graph/\` — Scoped call graph construction + DFS traversal
  - \`sync-detection/\` — Method call detection
  - \`interface-resolution/\` — Single-impl auto-resolve
  - \`detect-connections.ts\` — Orchestrator
- **Firm constraints:** §9.1.1 (no new packages), §9.1.2 (domain capability in extract-ts), §9.2.1 (ComponentIndex immutable value object)
- **Flexible decisions:** §9.2.2 (ScopedCallGraph implementation details), §9.2.3 (ExtractedLink extends Link)
- Use ts-morph for AST traversal (already a dependency of extract-ts)
- Follow existing patterns in \`domain/component-extraction/extractor.ts\` and \`domain/value-extraction/\`
- ExtractedLink extends Link from riviere-schema with optional \`_uncertain?: string\`
- Apply SD-004 (dependency injection), SD-001 (fail-fast), SD-010 (wrap primitives as value objects)
- Follow \`docs/conventions/codebase-structure.md\` CS-004 (organize by usage)

### Embedded Reasoning
**Why:** Components are identified and enriched. Detecting how they connect via method calls is the foundation of flow-based architecture extraction. Call graph + sync detection is the core capability everything else builds on.
**What:** ComponentIndex (immutable lookup), ScopedCallGraph (DFS traversal with type-based resolution), interface resolution (single-impl auto-resolve), sync connection detection (method calls → ExtractedLink[]).
**How:** Build ComponentIndex from EnrichedComponent[] + AST. For each component method, DFS through call sites resolving receiver types via declared types. Collapse non-component nodes. Produce one ExtractedLink per unique (source, target, type) tuple.

### Testing Strategy
- **Unit:** ComponentIndex construction and lookups (type-name, file:line). Call graph DFS traversal. Interface resolution (0, 1, N implementations). Sync detection for direct calls, transitive calls, chained calls.
- **Integration:** End-to-end from EnrichedComponent[] + ts-morph Project to ExtractedLink[] using multi-file test fixtures.
- **Edge Cases:** Empty inputs, cycles, dead ends, self-links, generic types, untyped variables, getters vs property access, await transparency.
- **Performance:** Not in scope for this task (D1.8).

### Dependencies
None — this is the foundational task.

### Verification
\`\`\`bash
pnpm nx test riviere-extract-ts -- --testNamePattern "connection-detection"
pnpm nx test riviere-extract-ts -- --coverage
\`\`\`

## Summary by CodeRabbit

* **New Features**
  * Added a connection-detection subsystem to analyze and emit component-to-component call links (multi‑hop tracing, interface resolution) and new public types/APIs for detection, indexing, extracted links, and structured errors.

* **Tests**
  * Added extensive unit and coverage tests for call-graph building, type resolution, interface resolution, tracing edge cases, deduplication, and overall detection behavior.

* **Chores**
  * Added a workspace dependency and updated TypeScript project references for the new subsystem.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

File: packages/riviere-extract-ts/src/domain/connection-detection/call-graph/build-call-graph.ts [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+    componentIndex,
+    component,
+    typeName,
+    calledMethodName,
+    currentCallSite,
+    rawLinks,
+    uncertainLinks,
+    options,
+  )
+}
```
[the remark this answers is not in the document]
Comment (NTCoding) [reply packages/riviere-extract-ts/src/domain/connection-detection/call-graph/build-call-graph.ts]: ✅ **Fixed**: Propagated options.strict to resolveInterface and resolveCallExpressionReceiverType. resolveTypeThroughInterface now returns {component, resolvedTypeName, uncertain}. Non-component implementations traced via resolvedTypeName. Uncertain links surfaced.

---
*Generated by Claude Code*

File: packages/riviere-extract-ts/src/domain/connection-detection/call-graph/deduplicate-links.spec.ts [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+    expect(result).toHaveLength(1)
+    expect(result[0]).toStrictEqual(
+      expect.objectContaining({
+        source: 'orders:useCase:Source',
+        target: '_unresolved',
+        _uncertain: "Receiver type is 'any'",
+      }),
+    )
+  })
```
[the remark this answers is not in the document]
Comment (NTCoding) [reply packages/riviere-extract-ts/src/domain/connection-detection/call-graph/deduplicate-links.spec.ts]: ✅ **Fixed**: Added tests for empty inputs, multiple uncertain links, and equal line numbers.

---
*Generated by Claude Code*

File: packages/riviere-extract-ts/src/domain/connection-detection/interface-resolution/resolve-interface.ts [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+}
+
+function createError(interfaceName: string, reason: string): ConnectionDetectionError {
+  return new ConnectionDetectionError({
+    file: '',
+    line: 0,
+    typeName: interfaceName,
+    reason,
+  })
+}
```
[the remark this answers is not in the document]
Comment (NTCoding) [reply packages/riviere-extract-ts/src/domain/connection-detection/interface-resolution/resolve-interface.ts]: ❌ **Rejected**: Interface resolution operates at project-level, not call-site-level. The interface name + reason provide sufficient debugging context. Call-site context is added by the caller (build-call-graph) when constructing the link.

---
*Generated by Claude Code*
